### PR TITLE
Add support for BSR <-> Strided conversion

### PR DESCRIPTION
**Pull request by cpuhrsch**
Extends 2D to ND support for BSR <-> Strided conversions via `to_sparse_bsr` and `to_dense`.

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 1]
```diff
@@ -676,13 +676,22 @@ Tensor empty_like_sparse_csr(
 }
 
 Tensor select_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/SparseCsrTensor.cpp:678]: NB: I still need to add a test for this

**Comment by cpuhrsch** [comment]
@pearu - I just added the tests for selecting batch dimensions of a BSR tensor. I'll rebase on top of https://github.com/pytorch/pytorch/pull/78359 once it's merged, which should resolve the CI failures. After that I'll do another pass on them.

File: aten/src/ATen/native/TensorConversions.cpp [hunk 2]
```diff
@@ -363,26 +363,57 @@ Tensor sparse_compressed_to_dense(
     return dst.add_(self);
   }
   if (self.layout() == kSparseBsr) {
-    TORCH_CHECK(self.dim() == 2, "Can only convert 2D SparseBsr to Strided.");
+    TORCH_CHECK(self.dim() == 2 || self.dim() == 3, "Can only convert 2D or 3D SparseBsr to Strided.");
+    auto crow_indices = self.crow_indices();
+    auto col_indices = self.col_indices();
+    if (self.dim() == 3) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorConversions.cpp:369]: We will want to add a runtime check here 
that verifies all the indices are the same. For now it's not possible to construct a batched BSR Tensor via `to_sparse_bsr` that violates this, but it's a risky assumption. I'll push a verification. Immediate follow up work includes generalizing this code to not depend on this assumption.

File: test/test_sparse_csr.py [hunk 3]
```diff
@@ -2209,8 +2241,91 @@ def test_dense_to_from_sparse_compressed(self, device, layout):
...
+        mask = torch.tensor([
+            [[True, True], [False, True]],
+            [[True, False], [True, True]]],
+            device=device).view((2, 2, 2, 1, 1))
+        mask = mask.expand((2, 2, 2, 2, 2))
+        mask = mask.transpose(2, 3)
+        mask = mask.reshape_as(dense)
+        dense = dense * mask
+        if layout == torch.sparse_bsr:
+            # this is not an error as long as the nse is equal for bsr
```
Comment (amjames) [inline test/test_sparse_csr.py:2302]: I want to highlight this for reviewers. 

The original 3-D extension by @cpuhrsch required that all batches have the same sparsity pattern, that is Both the number **and** location of specified blocks had to be the same. This is at odds with the blog post by @pearu on BSR invariants which states a more relaxed requirement.

> ...different batches of BSR tensors may have different sparsity patterns as long as the number of specified elements is the same for all batches.

I have implemented this with the more permissive requirement from the blog post, but it is fairly simple to reintroduce the strict check if needed.

Comment (cpuhrsch) [reply test/test_sparse_csr.py:2350]: @amjames - the additional constraints were only meant as a stepping stone to the more generic version that you now implemented :)

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 4]
```diff
@@ -821,25 +821,42 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
...
-            scalar_t *self_ip = result_ptr + self_i * result_stride;
-            *self_ip += *(source_ptr + i * source_stride) * alpha_value;
-        }
-      });
-    });
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+        ScalarType::Half,
+        ScalarType::Bool,
+        ScalarType::BFloat16,
+        ScalarType::ComplexHalf,
```
Comment (amjames) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:828]: I can cherry pick this commit and move it to a separate PR, but it was such a small change.

**Comment by amjames** [comment]
@cpuhrsch I resolved my issues with a very simple change. This is ready for a final review. If we want to move the `index_add` change out we will have to wait for that to merge this.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 4 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:828]: I know, but from a release process perspective it'll have to be a separate PR. Luckily GH1 makes this quick to merge. I'll stamp it as soon as it arrives. I'd also leave out the code formatting so it's truly just a single line change.

Comment (cpuhrsch) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:828]: I'd leave the change in this PR, create a separate, unstacked PR with this change only and then rebase once it's merged into master. That way you're not blocked and we don't have to change this current PR.

Comment (amjames) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:828]: Done, see #79897.

**Comment by amjames** [comment]
Superseded by #80354

**Pull request closed without merging**